ReactWindows/ReactNative.Net46/Bridge/DispatcherHelpers.cs

```diff
@@ -151,20 +151,82 @@ public static Task<T> CallOnDispatcher<T>(Dispatcher dispatcher, Func<T> func, b
         {
             if (allowInlining && IsOnDispatcher(dispatcher))
             {
-                return Task.FromResult(func());
+                try
+                {
+                    var result = func();
+                    return Task.FromResult(result);
+                }
+                catch (Exception e)
+                {
+                    return Task.FromException<T>(e);
+                }
             }
-            else
-            {
-                var taskCompletionSource = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
 
-                RunOnDispatcher(dispatcher, DispatcherPriority.Normal, () =>
+            var taskCompletionSource = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            RunOnDispatcher(dispatcher, DispatcherPriority.Normal, () =>
+            {
+                try
                 {
                     var result = func();
                     taskCompletionSource.SetResult(result);
+                }
+                catch (Exception e)
+                {
+                    taskCompletionSource.SetException(e);
+                }
+            });
+
+            return taskCompletionSource.Task;
+        }
+
+        /// <summary>
+        /// Invokes an async function on a specified dispatcher and asynchronously returns the
+        /// result.
+        /// </summary>
+        /// <param name="dispatcher">The Dispatcher to be used.</param>
+        /// <param name="func">The function to invoke.</param>
+        /// <param name="allowInlining">True if inlining is allowed when calling thread is on the same dispatcher as the one in the parameter.</param>
+        /// <returns>A task to await the result.</returns>
+        public static Task CallOnDispatcher(Dispatcher dispatcher, Func<Task> func, bool allowInlining = false)
```

Comment:
>CallOnDispatcher [](start = 27, length = 16)

Hmm. It's sort of a mistake that `DispatcherHelpers` is exposed as part of the public surface. Is this method signature being used in core somewhere? I think we have some spots where we `Unwrap` the result of a `CallOnDispatcher` call that returns Task...

Reply:
I.e., you could just as well implement this with:
```
CallOnDispatcher(dispatcher, func, allowInlining).Unwrap()
```

---
In reply to: [195436870](https://github.com/Microsoft/react-native-windows/pull/1867#discussion_r195436870) [](ancestors = 195436870)

Reply:
* if the API was really needed at all.

---
In reply to: [195437343](https://github.com/Microsoft/react-native-windows/pull/1867#discussion_r195437343) [](ancestors = 195437343,195436870)

Reply:
For the "DispatcherHelpers is public" part, pls don't remove it! Maybe its surface can be reduced but, at least on UWP, it is our only way to reliably dispatch to the main thread when in some obscure scenarios.
Basically the MainDispatcher can help with its cached during DispatcherHelpers.Initialize when trying to get CoreApplication.MainView or CoreApplication.MainView.Dispatcher throws ridiculous exceptions when executed from another thread.

---
In reply to: [195436870](https://github.com/Microsoft/react-native-windows/pull/1867#discussion_r195436870) [](ancestors = 195436870)

Reply:
Removed and will only have it in the fork.